### DISCUSSION TITLE
SQlite Database and Image Dock

### DISCUSSION THREAD

**Discussion by julas23** [brndnmtthws/conky#1618] [Ideas]
Hello everyone.

I´d like first of all, congrats for the work. I have been a conky user since 2009. And I really enjoy it.

I have 2 suggestion to make something really good, a little better.

I am now using SQLite3 to host all info I want to show up in Conky, hourly I run a script to check if something changed, and if does, update SQLite3. Could be something similar on Conky. All info hosted in SQLite3 database (file).

And now an important improvement. I see some issue to organize/align images as reference or information in conky.
I have a python calendar, and for example, when the month has more, or less, than 5 weeks, the output size changes and all the images came out of place. There is any way to hold or anchor the image in the specific line?

Follow below my conky conf.

conky.config = {
	background = true,
	update_interval = 1,
	total_run_times = 0,
	override_utf8_locale = true,
	double_buffer = true,
	no_buffers = true,
	text_buffer_size = 116384,
	max_user_text = 65536,
	temperature_unit = 'celsius',
	short_units = true,
	out_to_console = false,
    out_to_ncurses = false,
    out_to_stderr = false,
    out_to_x = true,
	own_window = true,
	own_window_type = 'panel', -- desktop, dock, panel, normal, override
	own_window_class = 'Conky',
	own_window_hints = 'below,undecorated,sticky,skip_taskbar,skip_pager',
	own_window_transparent = true,
	own_window_argb_visual = true,
	own_window_argb_value = 255,
	imlib_cache_size = 1000000,
	imlib_cache_flush_interval = 3600,
	border_width = 0,
	border_inner_margin = 0,
	border_outer_margin = 0,
	show_graph_scale = false,
	show_graph_range = false,
	minimum_width = 401,
	maximum_width = 401,
	minimum_height = 2160,
	xinerama_head = 0,
	net_avg_samples = 2,
    cpu_avg_samples = 2,
	alignment = 'top_right',
	extra_newline = false,
	uppercase = false,
	use_spacer = 'none',
	use_xft = true,
	xftalpha = 0,
	gap_x = 0,
	gap_y = 0,
	draw_shades = false,
	default_shade_color = '777777',
	draw_blended=true,
	draw_outline = false,
	draw_borders = false,
	draw_graph_borders = true,
	if_up_strictness = 'address',
	stippled_borders = 0,
	default_color = 'FFFFFF',
	default_outline_color = '777777',
	log_file = '~/.git/conky/config/conky_ju.log',
	color1 = 'FFFFFF',	-- white
	color2 = 'FFFF50',	-- yellow
	color3 = 'FFA800',	-- orage
	color4 = 'FF393B',	-- red
	color5 = '75839D',	-- lightblue
	color6 = '2C3548',	-- teal
	color7 = '7E1500',	-- darkred
	color8 = 'lightyellow',	-- lightyellow
	color9 = '335A29',	-- green
}

conky.text = [[
${goto 40}${color5}${font Terminus:Bold:size=10}Today${hr}${color}${font}

${voffset -32}${font Terminus:Bold:size=12}${goto 175}${color4}${execpi 10 /home/juliano/.git/conky/data weather_tmp_x}°C${color}${font}
${voffset -5}${font Terminus:Bold:size=12}${goto 175}${color1}${execpi 10 /home/juliano/.git/conky/data weather_tmp_n}°C${color}${font}
${voffset -6}${font Terminus:Bold:size=12}${goto 175}${color5}${execpi 10 /home/juliano/.git/conky/data weather_tmp_m}°C${color}${font}

${voffset -92}${font Terminus:Bold:size=25}${color1}${goto 230}${time %a %d %b}${color}${font}
${voffset -0}${font Terminus:Bold:size=34}${goto 230}${color2}${time %H:%M}${font Terminus:Bold:size=20}${alignr}${time %S}${color}${font}
${voffset -5}${font Terminus:Bold:size=12}${goto 232}New York:${alignr}${font Terminus:Regular:size=12}${tztime America/New_York %H:%M:%S}${font}
${voffset -10}${font Terminus:Bold:size=12}${goto 232}London:${alignr}${font Terminus:Regular:size=12}${tztime Europe/London %H:%M:%S}${font}
${voffset -10}${font Terminus:Bold:size=12}${goto 232}Tokyo:${alignr}${font Terminus:Regular:size=12}${tztime Asia/Tokyo %H:%M:%S}${font}

${voffset -45}${font Terminus:Bold:size=10}${color2}${goto 20}${execpi 10 ~/.git/conky/data next_1_day_fw}${goto 90}${execpi 10 ~/.git/conky/data next_2_day_fw}${goto 160}${execpi 10 ~/.git/conky/data next_3_day_fw}${color}${font}

${voffset -40}
${goto 40}${font Terminus:Bold:size=10}${color5}Calendar${hr}${color}${font}
${font Terminus:Bold:size=17}${alignc}${color}${execpi 10 ~/.git/conky/data calendar}${color}${font}

${voffset -60}
${goto 40}${font Terminus:Bold:size=10}${color5}System${hr}${color}

${goto 20}USER@HOST:${goto 120}${color3}${execpi 10 whoami}${color}${color4}@${color}${color5}${execpi 10 hostname -s}${color}
${goto 20}UPTIME:${goto 120}${color2}${uptime}${color}
${goto 20}DISTRO:${goto 120}${color2}${execpi 10 ~/.git/conky/data distroversion}${color}
${goto 20}KERNEL:${goto 120}${color2}${execpi 10 ~/.git/conky/data kernelversion}${color}
${goto 20}ENVIRONMENT:${goto 120}${color2}${execpi 10 ~/.git/conky/data graphic_envin}${color}
${goto 20}MANAGER:${goto 120}${color2}${execpi 10 ~/.git/conky/data envinrmanager}${color}
${goto 20}THEME:${goto 120}${color2}${execpi 10 ~/.git/conky/data envinrontheme}${color}
${goto 20}RESOLUTION:${goto 120}${color2}${execpi 10 ~/.git/conky/data x11resolution}${color}

${voffset -20}

${goto 40}${font Terminus:Bold:size=10}${color5}Hardware${hr}${color}

${goto 100}${color2}${execpi 10 ~/.git/conky/data mobo_manufact} ${execpi 10 ~/.git/conky/data mobo_modelver}${goto 240}${execpi 10 ~/.git/conky/data fan1_rpm_spee} RPM${alignr}${execpi 10 ~/.git/conky/data case_fan_rpms} RPM${color}

${goto 100}${color2}${execpi 10 ~/.git/conky/data gpu_manufactu}${color2}${alignr}${execpi 10 ~/.git/conky/data gpu_temperatu}°C${color}

${goto 100}${color2}${execpi 10 ~/.git/conky/data cpu_manufectu}${goto 240}${color}${color2}${cpu cpu0}%${color}${goto 275}${goto 295}${color2}${freq_g 0}GHz${color}${alignr}${color2}${execpi 10 ~/.git/conky/data cpu_temperatu}°C${color}
${alignr}${cpugraph cpu0 25,350 52ff00 6edd21 -t}

${voffset -30}

${goto 40}${color5}Memory${hr}${color}

${goto 20}RAM:${goto 100}${color2}${memmax}${goto 170}${color4}${mem}${color}${goto 250}${color9}${memfree}${color}${goto 330}${color2}${memperc}%${color}
${goto 100}${color2}${execpi 10 ~/.git/conky/data memory_slotty}${goto 170}${execpi 10 ~/.git/conky/data memory_cloksp}Mhz${color}${goto 250}${execpi 10 ~/.git/conky/data mobo_ddr_used} / ${execpi 10 ~/.git/conky/data mobo_ddr_slot} slots
${goto 20}${voffset -11}${alignr}${membar 10,50}
${goto 20}${voffset -32}${alignr}${memgraph 10,50 859900 DC322F scale -t}

${voffset -30}

${goto 40}${voffset 20}${color5}Network${hr}${color}

${goto 20} \
${if_existing /sys/class/net/enp7s0/operstate up}\
${if_existing /proc/net/route enp7s0}\
${goto 20}${image ~/.git/conky/img/g.png -p 365,755 -s 40x40}${goto 40}${color2}${addr enp7s0}${goto 195}${execpi 10 ~/.git/conky/data public_ip_add}${goto 345}On${color}\
${else}\
${goto 20}${image ~/.git/conky/img/r.png -p 365,755 -s 40x40}${goto 40}${color4}${addr enp7s0}${goto 195}Offline${goto 345}Off${color}\
${endif}\
${else}\
${goto 20}${image ~/.git/conky/img/r.png -p 365,755 -s 40x40}${goto 40}${color4}${addr enp7s0}${goto 195}Offline${goto 345}Off${color}\
${endif}
${goto 40}UL: ${upspeedgraph enp7s0 10,58 859900 DC322F scale -t}${goto 130} ${upspeed enp7s0} ${goto 195}DL:${goto 215} ${downspeedgraph enp7s0 10,58 859900 DC322F scale -t}${goto 285}${downspeed enp7s0}

${goto 20} \
${if_existing /sys/class/net/wlp6s0/operstate up}\
${if_existing /proc/net/route wlp6s0}\
${goto 20}${image ~/.git/conky/img/g.png -p 365,795 -s 40x40}${color2}${goto 40}${addr wlp6s0}${goto 195}${execpi 10 ~/.git/conky/data public_ip_add}${goto 315}${wireless_essid wlp6s0}${color}\
${else}\
${goto 20}${image ~/.git/conky/img/r.png -p 365,795 -s 40x40}${color4}${goto 40}${addr wlp6s0}${goto 195}Offline${goto 285}${wireless_essid wlp6s0}${color}\
${endif}\
${else}\
${goto 20}${image ~/.git/conky/img/r.png -p 365,795 -s 40x40}${color4}${goto 40}${addr wlp6s0}${goto 195}Offline${goto 285}${wireless_essid wlp6s0}${color}${color}\
${endif}
${goto 40}UL: ${upspeedgraph wlp6s0 10,58 859900 DC322F scale -t}${goto 130} ${upspeed wlp6s0} ${goto 195}DL:${goto 215} ${downspeedgraph wlp6s0 10,58 859900 DC322F scale -t}${goto 285} ${downspeed wlp6s0}${color2}${goto 347}${wireless_link_qual wlp6s0}${goto 355}%${color}

${voffset -30}

${goto 40}${voffset -5}${color5}Storage${hr}${color}

${goto 20}Sata:${color2}${goto 100}R: ${diskio_read /dev/sda}${goto 170}W: ${diskio_write /dev/sda}${goto 250}I: ${diskio /dev/sda}${color}${alignr}${diskiograph /dev/sda 10,50 859900 DC322F scale -t}
${goto 20}/ ${goto 100}${color2}${fs_size /}${color}${goto 170}${color4}${fs_used /}${goto 250}${color9}${fs_free /}${color}${alignr}${fs_bar 10,50 /}${color}

${goto 20}0: ${color2}${execpi 10 ~/.git/conky/data nvme0_tempera}°C${goto 100}R: ${diskio_read /dev/nvme0n1}${goto 170}W: ${diskio_write /dev/nvme0n1}${goto 250}I: ${diskio /dev/nvme0n1}${color}${alignr}${diskiograph /dev/nvme0n1 10,50 859900 DC322F scale -t}
${goto 20}/FS/GAME ${goto 100}${color2}${fs_size /FS/GAME}${color}${goto 170}${color4}${fs_used /FS/GAME}${goto 250}${color9}${fs_free /FS/GAME}${color}${alignr}${fs_bar 10,50 /FS/GAME}${color}

${goto 20}1: ${color2}${execpi 10 ~/.git/conky/data nvme1_tempera}°C${goto 100}R: ${diskio_read /dev/nvme1n1}${goto 170}W: ${diskio_write /dev/nvme1n1}${goto 250}I: ${diskio /dev/nvme1n1}${color}${alignr}${diskiograph /dev/nvme1n1 10,50 859900 DC322F scale -t}
${goto 20}2: ${color2}${execpi 10 ~/.git/conky/data nvme2_tempera}°C${goto 100}R: ${diskio_read /dev/nvme2n1}${goto 170}W: ${diskio_write /dev/nvme2n1}${goto 250}I: ${diskio /dev/nvme2n1}${color}${alignr}${diskiograph /dev/nvme2n1 10,50 859900 DC322F scale -t}
${goto 20}/FS/DATA ${goto 100}${color2}${fs_size /FS/DATA}${color}${goto 170}${color4}${fs_used /FS/DATA}${goto 250}${color9}${fs_free /FS/DATA}${color}${alignr}${fs_bar 10,50 /FS/DATA}${color}
${if_mounted /FS/BACK}
${goto 20}U: ${color2}${goto 100}R: ${diskio_read /dev/sda1}${goto 170}W: ${diskio_write /dev/sda1}${goto 250}I: ${diskio /dev/sda1}${color}${alignr}${diskiograph /dev/sda1 10,50 859900 DC322F scale -t}
${goto 20}/FS/BACK ${goto 100}${color2}${fs_size /FS/BACK}${color}${goto 170}${color4}${fs_used /FS/BACK}${goto 250}${color9}${fs_free /FS/BACK}${color}${alignr}${fs_bar 10,50 /FS/BACK}${color}
${endif}

${voffset -20}

${goto 40}${voffset -20}${color5}CPU Usage${hr}${color}${font Terminus:Bold:size=7}

${color7}${goto 5}${top pid 1}${goto 85}${top name 1}${goto 230}${top cpu 1}${goto 340}${top user 1}
${color4}${goto 5}${top pid 2}${goto 85}${top name 2}${goto 230}${top cpu 2}${goto 340}${top user 2}
${color orange}${goto 5}${top pid 3}${goto 85}${top name 3}${goto 230}${top cpu 3}${goto 340}${top user 3}
${color2}${goto 5}${top pid 4}${goto 85}${top name 4}${goto 230}${top cpu 4}${goto 340}${top user 4}
${color8}${goto 5}${top pid 5}${goto 85}${top name 5}${goto 230}${top cpu 5}${goto 340}${top user 5}
${color white}${goto 5}${top pid 6}${goto 85}${top name 6}${goto 230}${top cpu 6}${goto 340}${top user 6}

${color7}${goto 5}${top_mem pid 1}${goto 85}${top_mem name 1}${goto 247}${top mem_res 1}${goto 340}${top mem_vsize 1}
${color4}${goto 5}${top_mem pid 2}${goto 85}${top_mem name 2}${goto 247}${top mem_res 2}${goto 340}${top mem_vsize 2}
${color orange}${goto 5}${top_mem pid 3}${goto 85}${top_mem name 3}${goto 247}${top mem_res 3}${goto 340}${top mem_vsize 3}
${color2}${goto 5}${top_mem pid 4}${goto 85}${top_mem name 4}${goto 247}${top mem_res 4}${goto 340}${top mem_vsize 4}
${color8}${goto 5}${top_mem pid 5}${goto 85}${top_mem name 5}${goto 247}${top mem_res 5}${goto 340}${top mem_vsize 5}
${color white}${goto 5}${top_mem pid 6}${goto 85}${top_mem name 6}${goto 247}${top mem_res 6}${goto 340}${top mem_vsize 6}
${color}FG = $running_processes${alignc}BG = $processes${alignr}TH = $threads

${voffset -10}

${goto 40}${font Mono:Regular:size=8}${color5}Notes${hr}${color}
${voffset 10}${font Mono:Regular:size=6}${exec ~/.git/conky/data note}${font}
${voffset -20}
${goto 40}${font DeJavu:Regular:size=8}${color5}Tasks${hr}${color}${font}
${voffset -10}${font Mono:Regular:size=6}${exec ~/.git/conky/data task}${font}

${if_match "${execpi 10 ~/.git/conky/data chk_users_log && echo 1 || echo 0}" == "1"}${image ~/.git/conky/img/x.png -p 60,550 -s 300x600}${endif}
${image ~/.git/conky/img/u.png -p 155,23 -s 15x15 -f 86400 -n 1}
${image ~/.git/conky/img/t.png -p 150,42 -s 20x20 -f 86400 -n 1}
${image ~/.git/conky/img/d.png -p 155,67 -s 15x15 -f 86400 -n 1}
${execpi 10 brweather --imagem1=CN -p 0,-20 -s 150x150 -f 3600}
${execpi 10 brweather --imagem1=C1 -p 20,95 -s 50x50 -f 3600}
${execpi 10 brweather --imagem1=C2 -p 90,95 -s 50x50 -f 3600}
${execpi 10 brweather --imagem1=C3 -p 160,95 -s 50x50 -f 3600}

${image ~/.git/conky/img/pin.png -p 0,0 -s 35x35 -f 86400 -n 1}
${image ~/.git/conky/img/cal.png -p 0,157 -s 40x40 -f 86400 -n 1}
${image ~/.git/conky/img/yoda.ico -p -5,365 -s 50x50 -f 86400 -n 1}
${image ~/.git/conky/img/arch.png -p 265,410 -s 130x95 -f 86400 -n 1}
${image ~/.git/conky/img/pc.png -p 0,530 -s 30x30 -f 86400 -n 1}
${image ~/.git/conky/img/mobo.png -p 48,555 -s 25x25 -f 86400 -n 1}
${image ~/.git/conky/img/fan.png -p 210,553 -s 30x30 -f 86400 -n 1}
${image ~/.git/conky/img/fan.png -p 315,553 -s 30x30 -f 86400 -n 1}
${image ~/.git/conky/img/gpu.png -p 45,582 -s 40x26 -f 86400 -n 1}
${image ~/.git/conky/img/term.png -p 358,583 -s 20x20 -f 86400 -n 1}
${image ~/.git/conky/img/cpu.png -p 45,605 -s 30x30 -f 86400 -n 1}
${image ~/.git/conky/img/term.png -p 358,610 -s 20x20 -f 86400 -n 1}
${image ~/.git/conky/img/ram.png -p 0,665 -s 30x30 -f 86400 -n 1}
${image ~/.git/conky/img/network.png -p 0,735 -s 30x30 -f 86400 -n 1}
${image ~/.git/conky/img/cable.png -p 5,770 -s 25x25 -f 86400 -n 1}
${image ~/.git/conky/img/wifi.png -p 0,810 -s 30x20 -f 86400 -n 1}
${image ~/.git/conky/img/nvme.png -p 0,845 -s 40x20 -f 86400 -n 1}
${image ~/.git/conky/img/tux.png -p 312,875 -s 25x32 -f 86400 -n 1}
${image ~/.git/conky/img/joypad.png -p 305,912 -s 35x35 -f 86400 -n 1}
${image ~/.git/conky/img/raid.png -p 307,955 -s 35x35 -f 86400 -n 1}
${image ~/.git/conky/img/backup.png -p 310,1005 -s 35x35 -f 86400 -n 1}
${image ~/.git/conky/img/cpusage.png -p 0,1035 -s 28x28 -f 86400 -n 1}
${image ~/.git/conky/img/notes.png -p 0,1215 -s 30x30 -f 86400 -n 1}
${image ~/.git/conky/img/todo.png -p 0,1352 -s 30x30 -f 86400 -n 1}
]]

![1](https://github.com/brndnmtthws/conky/assets/92172251/206832f5-b258-4898-b056-e2295ebd335d)

**Comment by Caellian** [comment]
With regards to the calendar issue - you can use Cairo from Lua to draw the calendar and manage your layout programmatically, see the "Using Lua scripts" section of the [wiki](https://github.com/brndnmtthws/conky/wiki).

As for SQLite - you mean use SQLite as an output target or making Conky be able to read from a db file?

**Reply by julas23** [reply]
I mean to use a SQLite Database for conky data, and grabbing from there for example.

**Reply by julas23** [reply]
To avoid keep collecting everything again every refresh

**Reply by Caellian** [reply]
You can combine Lua's `io.popen` and a [cmd query](https://sqlite.org/cli.html#using_sqlite3_in_a_shell_script) to achieve that.

**Reply by Caellian** [reply]
There _is_ MySQL support in conky, so you could add a SQLite as well if you have time, or just modify it locally.
